component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/EmailOtpServiceImpl.java

```diff
@@ -548,4 +571,213 @@ private int getTenantId() {
+
+        Map<String, String> updatedClaims = new HashMap<>();
+        if ((currentAttempts + 1) >= maxAttempts) {
+            // Calculate the incremental unlock time interval in milli seconds.
+            unlockTimePropertyValue = (long) (unlockTimePropertyValue * 1000 * 60 * Math.pow(unlockTimeRatio,
+                    failedLoginLockoutCountValue));
+            populateAccountLockClaims(unlockTimePropertyValue, failedLoginLockoutCountValue, updatedClaims);
+            setUserClaimValues(user, updatedClaims);
+            FailureReasonDTO error = showFailureReason
+                    ? new FailureReasonDTO(Constants.ErrorMessage.CLIENT_ACCOUNT_LOCKED, userId)
+                    : null;
+            return new ValidationResponseDTO(userId, false, error);
+        } else {
+            updatedClaims.put(Constants.EMAIL_OTP_FAILED_ATTEMPTS_CLAIM, String.valueOf(currentAttempts + 1));
+            setUserClaimValues(user, updatedClaims);
+            return null;
+        }
+    }
+
+    private ValidationResponseDTO createAccountLockedResponse(String userId, boolean showFailureReason) {
+
+        FailureReasonDTO error = showFailureReason
+                ? new FailureReasonDTO(Constants.ErrorMessage.CLIENT_ACCOUNT_LOCKED, userId)
+                : null;
+        return new ValidationResponseDTO(userId, false, error);
+    }
+
+    private int getCurrentAttempts(Map<String, String> claimValues) {
+
+        if (NumberUtils.isNumber(claimValues.get(Constants.EMAIL_OTP_FAILED_ATTEMPTS_CLAIM))) {
+            return Integer.parseInt(claimValues.get(Constants.EMAIL_OTP_FAILED_ATTEMPTS_CLAIM));
+        }
+        return 0;
+    }
+
+    private int getFailedLoginLockoutCount(Map<String, String> claimValues) {
+
+        if (NumberUtils.isNumber(claimValues.get(Constants.FAILED_LOGIN_LOCKOUT_COUNT_CLAIM))) {
+            return Integer.parseInt(claimValues.get(Constants.FAILED_LOGIN_LOCKOUT_COUNT_CLAIM));
+        }
+        return 0;
+    }
+
+    private void populateAccountLockClaims(long unlockTimePropertyValue, int failedLoginLockoutCountValue,
+                                           Map<String, String> updatedClaims) {
+
+        // Calculate unlock time by adding current time and unlock time interval in milliseconds.
+        long unlockTime = System.currentTimeMillis() + unlockTimePropertyValue;
+        updatedClaims.put(Constants.ACCOUNT_LOCKED_CLAIM, Boolean.TRUE.toString());
+        updatedClaims.put(Constants.EMAIL_OTP_FAILED_ATTEMPTS_CLAIM, "0");
+        updatedClaims.put(Constants.ACCOUNT_UNLOCK_TIME_CLAIM, String.valueOf(unlockTime));
+        updatedClaims.put(Constants.FAILED_LOGIN_LOCKOUT_COUNT_CLAIM,
+                String.valueOf(failedLoginLockoutCountValue + 1));
+        updatedClaims.put(Constants.ACCOUNT_LOCKED_REASON_CLAIM_URI, Constants.MAX_EMAIL_OTP_ATTEMPTS_EXCEEDED);
+        IdentityUtil.threadLocalProperties.get().put(Constants.ADMIN_INITIATED, false);
```

Comment:
Why do we need a thread-local here? does it get properly cleared out?

Reply:
this is required in the account lock handler level to determine the correct email template to be used.